// @massalabs/massa-web3/dist/esm/index.js
/** Polyfills */
import { Buffer } from 'buffer';
import EventEmitter from 'events';
// Check if we are on browser
if (typeof window !== 'undefined') {
    window.Buffer = Buffer;
    window.EventEmitter = EventEmitter;
}
export { ProviderType } from './interfaces/IProvider';
export { EOperationStatus } from './interfaces/EOperationStatus';
export { OperationTypeId, } from './interfaces/OperationTypes';
/** Exposed clients and factories */
export { ClientFactory, DefaultProviderUrls } from './web3/ClientFactory';
export { Client } from './web3/Client';
export { PublicApiClient } from './web3/PublicApiClient';
export { PrivateApiClient } from './web3/PrivateApiClient';
export { WalletClient } from './web3/WalletClient';
export { EventPoller, ON_MASSA_EVENT_DATA, ON_MASSA_EVENT_ERROR, } from './web3/EventPoller';
export { SmartContractsClient, MASSA_PROTOFILE_KEY, PROTO_FILE_SEPARATOR, } from './web3/SmartContractsClient';
export { withTimeoutRejection } from './utils/time';
export { Web3Account } from './web3/accounts/Web3Account';
export { WalletProviderAccount } from './web3/accounts/WalletProviderAccount';
/**
 * This namespace provides utility functions, such as unit conversion, serialization, encoding, etc.
 */
export * as utils from './utils';
export * from './utils/keyAndAddresses';
export * from '@massalabs/web3-utils';
