// @massalabs/massa-sc-deployer/build/deployer.js
import * as __import0 from "massa";
async function instantiate(module, imports = {}) {
  const __module0 = imports.massa;
  const adaptedImports = {
    env: Object.assign(Object.create(globalThis), imports.env || {}, {
      abort(message, fileName, lineNumber, columnNumber) {
        // ~lib/builtins/abort(~lib/string/String | null?, ~lib/string/String | null?, u32?, u32?) => void
        message = __liftString(message >>> 0);
        fileName = __liftString(fileName >>> 0);
        lineNumber = lineNumber >>> 0;
        columnNumber = columnNumber >>> 0;
        (() => {
          // @external.js
          throw Error(`${message} in ${fileName}:${lineNumber}:${columnNumber}`);
        })();
      },
    }),
    massa: Object.assign(Object.create(__module0), {
      assembly_script_has_op_key(key) {
        // ~lib/@massalabs/massa-as-sdk/assembly/env/env/env.hasOpKey(~lib/staticarray/StaticArray<u8>) => ~lib/staticarray/StaticArray<u8>
        key = __liftStaticArray(pointer => new Uint8Array(memory.buffer)[pointer >>> 0], 0, key >>> 0);
        return __lowerStaticArray((pointer, value) => { new Uint8Array(memory.buffer)[pointer >>> 0] = value; }, 5, 0, __module0.assembly_script_has_op_key(key), Uint8Array) || __notnull();
      },
      assembly_script_get_op_data(key) {
        // ~lib/@massalabs/massa-as-sdk/assembly/env/env/env.getOpData(~lib/staticarray/StaticArray<u8>) => ~lib/staticarray/StaticArray<u8>
        key = __liftStaticArray(pointer => new Uint8Array(memory.buffer)[pointer >>> 0], 0, key >>> 0);
        return __lowerStaticArray((pointer, value) => { new Uint8Array(memory.buffer)[pointer >>> 0] = value; }, 5, 0, __module0.assembly_script_get_op_data(key), Uint8Array) || __notnull();
      },
      assembly_script_create_sc(bytecode) {
        // ~lib/@massalabs/massa-as-sdk/assembly/env/env/env.createSC(~lib/staticarray/StaticArray<u8>) => ~lib/string/String
        bytecode = __liftStaticArray(pointer => new Uint8Array(memory.buffer)[pointer >>> 0], 0, bytecode >>> 0);
        return __lowerString(__module0.assembly_script_create_sc(bytecode)) || __notnull();
      },
      assembly_script_function_exists(address, func) {
        // ~lib/@massalabs/massa-as-sdk/assembly/env/env/env.functionExists(~lib/string/String, ~lib/string/String) => bool
        address = __liftString(address >>> 0);
        func = __liftString(func >>> 0);
        return __module0.assembly_script_function_exists(address, func) ? 1 : 0;
      },
      assembly_script_call(address, func, param, coins) {
        // ~lib/@massalabs/massa-as-sdk/assembly/env/env/env.call(~lib/string/String, ~lib/string/String, ~lib/staticarray/StaticArray<u8>, u64) => ~lib/staticarray/StaticArray<u8>
        address = __liftString(address >>> 0);
        func = __liftString(func >>> 0);
        param = __liftStaticArray(pointer => new Uint8Array(memory.buffer)[pointer >>> 0], 0, param >>> 0);
        coins = BigInt.asUintN(64, coins);
        return __lowerStaticArray((pointer, value) => { new Uint8Array(memory.buffer)[pointer >>> 0] = value; }, 5, 0, __module0.assembly_script_call(address, func, param, coins), Uint8Array) || __notnull();
      },
      assembly_script_generate_event(event) {
        // ~lib/@massalabs/massa-as-sdk/assembly/env/env/env.generateEvent(~lib/string/String) => void
        event = __liftString(event >>> 0);
        __module0.assembly_script_generate_event(event);
      },
    }),
  };
  const { exports } = await WebAssembly.instantiate(module, adaptedImports);
  const memory = exports.memory || imports.env.memory;
  const adaptedExports = Object.setPrototypeOf({
    main(_) {
      // assembly/contracts/deployer/main(~lib/staticarray/StaticArray<u8>) => void
      _ = __lowerStaticArray((pointer, value) => { new Uint8Array(memory.buffer)[pointer >>> 0] = value; }, 5, 0, _, Uint8Array) || __notnull();
      exports.main(_);
    },
  }, exports);
  function __liftString(pointer) {
    if (!pointer) return null;
    const
      end = pointer + new Uint32Array(memory.buffer)[pointer - 4 >>> 2] >>> 1,
      memoryU16 = new Uint16Array(memory.buffer);
    let
      start = pointer >>> 1,
      string = "";
    while (end - start > 1024) string += String.fromCharCode(...memoryU16.subarray(start, start += 1024));
    return string + String.fromCharCode(...memoryU16.subarray(start, end));
  }
  function __lowerString(value) {
    if (value == null) return 0;
    const
      length = value.length,
      pointer = exports.__new(length << 1, 2) >>> 0,
      memoryU16 = new Uint16Array(memory.buffer);
    for (let i = 0; i < length; ++i) memoryU16[(pointer >>> 1) + i] = value.charCodeAt(i);
    return pointer;
  }
  function __liftStaticArray(liftElement, align, pointer) {
    if (!pointer) return null;
    const
      length = new Uint32Array(memory.buffer)[pointer - 4 >>> 2] >>> align,
      values = new Array(length);
    for (let i = 0; i < length; ++i) values[i] = liftElement(pointer + (i << align >>> 0));
    return values;
  }
  function __lowerStaticArray(lowerElement, id, align, values, typedConstructor) {
    if (values == null) return 0;
    const
      length = values.length,
      buffer = exports.__pin(exports.__new(length << align, id)) >>> 0;
    if (typedConstructor) {
      new typedConstructor(memory.buffer, buffer, length).set(values);
    } else {
      for (let i = 0; i < length; i++) lowerElement(buffer + (i << align >>> 0), values[i]);
    }
    exports.__unpin(buffer);
    return buffer;
  }
  function __notnull() {
    throw TypeError("value must not be null");
  }
  return adaptedExports;
}
export const {
  memory,
  main
} = await (async url => instantiate(
  await (async () => {
    try { return await globalThis.WebAssembly.compileStreaming(globalThis.fetch(url)); }
    catch { return globalThis.WebAssembly.compile(await (await import("node:fs/promises")).readFile(url)); }
  })(), {
    massa: __maybeDefault(__import0),
  }
))(new URL("deployer.wasm", import.meta.url));
function __maybeDefault(module) {
  return typeof module.default === "object" && Object.keys(module).length == 1
    ? module.default
    : module;
}
