// @massalabs/wallet-provider/dist/cmd/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.disconnectBearby = exports.connectBearby = exports.MassaStationAccount = exports.Provider = exports.EAccountImportResponse = exports.EAccountDeletionResponse = exports.Account = exports.getProviderByName = exports.registerProvider = exports.providers = exports.AvailableCommands = void 0;
/** Polyfills */
const buffer_1 = require("buffer");
// Check if we are on browser
if (typeof window !== 'undefined') {
    window.Buffer = buffer_1.Buffer;
}
const Connector_1 = require("./connector/Connector");
const Provider_1 = require("./provider/Provider");
const MassaStationProvider_1 = require("./massaStation/MassaStationProvider");
const BearbyConnect_1 = require("./bearbyWallet/BearbyConnect");
const BearbyProvider_1 = require("./bearbyWallet/BearbyProvider");
const time_1 = require("./utils/time");
var AvailableCommands;
(function (AvailableCommands) {
    AvailableCommands["ProviderListAccounts"] = "LIST_ACCOUNTS";
    AvailableCommands["ProviderDeleteAccount"] = "DELETE_ACCOUNT";
    AvailableCommands["ProviderImportAccount"] = "IMPORT_ACCOUNT";
    AvailableCommands["ProviderGetNodesUrls"] = "GET_NODES_URLS";
    AvailableCommands["ProviderGetNetwork"] = "GET_NETWORK";
    AvailableCommands["AccountBalance"] = "ACCOUNT_BALANCE";
    AvailableCommands["AccountSign"] = "ACCOUNT_SIGN";
    AvailableCommands["ProviderGenerateNewAccount"] = "GENERATE_NEW_ACCOUNT";
    AvailableCommands["AccountSellRolls"] = "ACCOUNT_SELL_ROLLS";
    AvailableCommands["AccountBuyRolls"] = "ACCOUNT_BUY_ROLLS";
    AvailableCommands["AccountSendTransaction"] = "ACCOUNT_SEND_TRANSACTION";
    AvailableCommands["AccountCallSC"] = "ACCOUNT_CALL_SC";
})(AvailableCommands = exports.AvailableCommands || (exports.AvailableCommands = {}));
/**
 * Get the list of providers that are available to interact with.
 *
 * @param retry - If true, will retry to get the list of providers if none are available.
 * @param pollInterval - The timeout in milliseconds to wait between retries. default is 2000ms.
 * @param timeout - The timeout in milliseconds to wait before giving up. default is 3000ms.
 *
 * @returns An array of providers.
 */
async function providers(retry = true, timeout = 3000, pollInterval = 500) {
    const startTime = Date.now();
    await Connector_1.connector.startMassaStationDiscovery();
    let bearby;
    if (await (0, BearbyConnect_1.detectBearby)()) {
        bearby = new BearbyProvider_1.BearbyProvider('BEARBY');
    }
    while (Date.now() - startTime < timeout) {
        const providerInstances = getProviderInstances();
        if (bearby)
            providerInstances.push(bearby);
        if (!retry || providerInstances.length > 0) {
            return providerInstances;
        }
        await (0, time_1.wait)(pollInterval);
    }
    return [];
}
exports.providers = providers;
function getProviderInstances() {
    const availableProviders = Object.keys(Connector_1.connector.getWalletProviders());
    const providerInstances = availableProviders.map((providerName) => {
        if (providerName === MassaStationProvider_1.MASSA_STATION_PROVIDER_NAME) {
            return new MassaStationProvider_1.MassaStationProvider(Connector_1.connector.getProviderInfo(providerName));
        }
        else {
            return new Provider_1.Provider(providerName);
        }
    });
    return providerInstances;
}
/**
 * Manually register a provider to interact with.
 *
 * @param name - The name of the provider.
 * @param id - The id of the HTML element that is used to communicate with the provider.
 */
function registerProvider(name, id = Connector_1.MASSA_WINDOW_OBJECT) {
    if (typeof document !== 'undefined') {
        const registerEvent = new CustomEvent('register', {
            detail: { providerName: name },
        });
        const element = document?.getElementById(id);
        if (element) {
            element.dispatchEvent(registerEvent);
        }
    }
}
exports.registerProvider = registerProvider;
async function getProviderByName(providerName) {
    const providersList = await providers();
    return providersList.find((p) => p.name() === providerName);
}
exports.getProviderByName = getProviderByName;
var account_1 = require("./account");
Object.defineProperty(exports, "Account", { enumerable: true, get: function () { return account_1.Account; } });
var provider_1 = require("./provider");
Object.defineProperty(exports, "EAccountDeletionResponse", { enumerable: true, get: function () { return provider_1.EAccountDeletionResponse; } });
Object.defineProperty(exports, "EAccountImportResponse", { enumerable: true, get: function () { return provider_1.EAccountImportResponse; } });
Object.defineProperty(exports, "Provider", { enumerable: true, get: function () { return provider_1.Provider; } });
var MassaStationAccount_1 = require("./massaStation/MassaStationAccount");
Object.defineProperty(exports, "MassaStationAccount", { enumerable: true, get: function () { return MassaStationAccount_1.MassaStationAccount; } });
var BearbyConnect_2 = require("./bearbyWallet/BearbyConnect");
Object.defineProperty(exports, "connectBearby", { enumerable: true, get: function () { return BearbyConnect_2.connectBearby; } });
Object.defineProperty(exports, "disconnectBearby", { enumerable: true, get: function () { return BearbyConnect_2.disconnectBearby; } });
