// @massalabs/massa-web3/dist/cmd/web3/accounts/WalletProviderAccount.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WalletProviderAccount = void 0;
const web3_utils_1 = require("@massalabs/web3-utils");
class WalletProviderAccount {
    account;
    constructor(account) {
        this.account = account;
    }
    async sign(data) {
        if (data instanceof Uint8Array) {
            data = Buffer.from(data);
        }
        const signatureData = await this.account.sign(data);
        return {
            publicKey: signatureData.publicKey,
            base58Encoded: signatureData.base58Encoded,
        };
    }
    address() {
        return this.account.address();
    }
    async sellRolls(rollsData) {
        let res = await this.account.sellRolls(rollsData.amount, rollsData.fee);
        return res.operationId;
    }
    async buyRolls(rollsData) {
        let res = await this.account.buyRolls(rollsData.amount, rollsData.fee);
        return res.operationId;
    }
    async sendTransaction(txData) {
        let res = await this.account.sendTransaction(txData.amount, txData.recipientAddress, txData.fee);
        return res.operationId;
    }
    async callSmartContract(callData) {
        let params = callData.parameter;
        let paramToSend;
        if (params instanceof Array) {
            paramToSend = new Uint8Array(params);
        }
        if (params instanceof web3_utils_1.Args) {
            paramToSend = params;
        }
        let res = (await this.account.callSC(callData.targetAddress, callData.functionName, paramToSend, callData.coins, callData.fee, callData.maxGas, false));
        return res.operationId;
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    async deploySmartContract(_) {
        throw new Error('deploySmartContract with an account from WalletProvider is not implemented yet.');
    }
    async verify() {
        return;
    }
}
exports.WalletProviderAccount = WalletProviderAccount;
