// @massalabs/web3-utils/dist/cmd/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CHAIN_ID_DNS_ADDRESS_MAP = exports.CHAIN_ID_RPC_URL_MAP = exports.DefaultProviderUrls = exports.CHAIN_ID = exports.CHAIN_ID_TO_NETWORK_NAME = exports.SANDBOX_CHAIN_ID = exports.LABNET_CHAIN_ID = exports.SECURENET_CHAIN_ID = exports.BUILDNET_CHAIN_ID = exports.MAINNET_CHAIN_ID = exports.SANDBOX = exports.LABNET = exports.SECURENET = exports.BUILDNET = exports.MAINNET = exports.MIN_GAS_CALL = exports.MAX_GAS_CALL = exports.MAX_GAS_DEPLOYMENT = exports.MAX_GAS_EXECUTE_SC = exports.BASE_ACCOUNT_CREATION_COST = exports.NEW_LEDGER_ENTRY_COST = exports.STORAGE_BYTE_COST = exports.KEYS_VERSION_NUMBER = exports.ADDRESS_PREFIX_LENGTH = exports.ADDRESS_CONTRACT_PREFIX = exports.ADDRESS_USER_PREFIX = exports.PUBLIC_KEY_PREFIX = exports.SECRET_KEY_PREFIX = void 0;
const converters_1 = require("./converters");
/* -------------------------------------------------------------------------- */
/*                                    KEYS                                    */
/* -------------------------------------------------------------------------- */
exports.SECRET_KEY_PREFIX = 'S';
exports.PUBLIC_KEY_PREFIX = 'P';
exports.ADDRESS_USER_PREFIX = 'AU';
exports.ADDRESS_CONTRACT_PREFIX = 'AS';
exports.ADDRESS_PREFIX_LENGTH = 2;
exports.KEYS_VERSION_NUMBER = 0;
/* -------------------------------------------------------------------------- */
/*                                STORAGE COST                                */
/* -------------------------------------------------------------------------- */
exports.STORAGE_BYTE_COST = (0, converters_1.fromMAS)(0.0001);
exports.NEW_LEDGER_ENTRY_COST = exports.STORAGE_BYTE_COST * 4n;
exports.BASE_ACCOUNT_CREATION_COST = (0, converters_1.fromMAS)(0.001);
/* -------------------------------------------------------------------------- */
/*                                 GAS LIMIT                                 */
/* -------------------------------------------------------------------------- */
exports.MAX_GAS_EXECUTE_SC = 3980167295n;
exports.MAX_GAS_DEPLOYMENT = exports.MAX_GAS_EXECUTE_SC;
exports.MAX_GAS_CALL = 4294167295n;
exports.MIN_GAS_CALL = 2100000n;
/* -------------------------------------------------------------------------- */
/*                                   NETWORK                                  */
/* -------------------------------------------------------------------------- */
exports.MAINNET = 'MainNet';
exports.BUILDNET = 'BuildNet';
exports.SECURENET = 'SecureNet';
exports.LABNET = 'LabNet';
exports.SANDBOX = 'Sandbox';
exports.MAINNET_CHAIN_ID = 77658377n;
exports.BUILDNET_CHAIN_ID = 77658366n;
exports.SECURENET_CHAIN_ID = 77658383n;
exports.LABNET_CHAIN_ID = 77658376n;
exports.SANDBOX_CHAIN_ID = 77n;
// Adjusted: Use the values of the constants as keys
exports.CHAIN_ID_TO_NETWORK_NAME = {
    [exports.MAINNET_CHAIN_ID.toString()]: exports.MAINNET,
    [exports.BUILDNET_CHAIN_ID.toString()]: exports.BUILDNET,
    [exports.SECURENET_CHAIN_ID.toString()]: exports.SECURENET,
    [exports.LABNET_CHAIN_ID.toString()]: exports.LABNET,
    [exports.SANDBOX_CHAIN_ID.toString()]: exports.SANDBOX,
}; // type is inferred as the specific, unchangeable structure
// NETWORK_NAME_TO_CHAIN_ID mapping remains the same
exports.CHAIN_ID = {
    [exports.MAINNET]: exports.MAINNET_CHAIN_ID,
    [exports.BUILDNET]: exports.BUILDNET_CHAIN_ID,
    [exports.SECURENET]: exports.SECURENET_CHAIN_ID,
    [exports.LABNET]: exports.LABNET_CHAIN_ID,
    [exports.SANDBOX]: exports.SANDBOX_CHAIN_ID,
}; // type is inferred as the specific, unchangeable structure
var DefaultProviderUrls;
(function (DefaultProviderUrls) {
    DefaultProviderUrls["MAINNET"] = "https://mainnet.massa.net/api/v2";
    DefaultProviderUrls["TESTNET"] = "https://test.massa.net/api/v2";
    DefaultProviderUrls["BUILDNET"] = "https://buildnet.massa.net/api/v2";
    DefaultProviderUrls["LABNET"] = "https://labnet.massa.net/api/v2";
    DefaultProviderUrls["LOCALNET"] = "http://127.0.0.1";
})(DefaultProviderUrls = exports.DefaultProviderUrls || (exports.DefaultProviderUrls = {}));
exports.CHAIN_ID_RPC_URL_MAP = {
    [exports.MAINNET_CHAIN_ID.toString()]: DefaultProviderUrls.MAINNET,
    [exports.BUILDNET_CHAIN_ID.toString()]: DefaultProviderUrls.BUILDNET,
    [exports.SANDBOX_CHAIN_ID.toString()]: DefaultProviderUrls.LOCALNET,
}; // type is inferred as the specific, unchangeable structure
exports.CHAIN_ID_DNS_ADDRESS_MAP = {
    [exports.MAINNET_CHAIN_ID.toString()]: 'AS1q5hUfxLXNXLKsYQVXZLK7MPUZcWaNZZsK7e9QzqhGdAgLpUGT',
    [exports.BUILDNET_CHAIN_ID.toString()]: 'AS12qKAVjU1nr66JSkQ6N4Lqu4iwuVc6rAbRTrxFoynPrPdP1sj3G',
}; // type is inferred as the specific, unchangeable structure
