// @massalabs/web3-utils/dist/esm/constants.js
import { fromMAS } from './converters';
/* -------------------------------------------------------------------------- */
/*                                    KEYS                                    */
/* -------------------------------------------------------------------------- */
export const SECRET_KEY_PREFIX = 'S';
export const PUBLIC_KEY_PREFIX = 'P';
export const ADDRESS_USER_PREFIX = 'AU';
export const ADDRESS_CONTRACT_PREFIX = 'AS';
export const ADDRESS_PREFIX_LENGTH = 2;
export const KEYS_VERSION_NUMBER = 0;
/* -------------------------------------------------------------------------- */
/*                                STORAGE COST                                */
/* -------------------------------------------------------------------------- */
export const STORAGE_BYTE_COST = fromMAS(0.0001);
export const NEW_LEDGER_ENTRY_COST = STORAGE_BYTE_COST * 4n;
export const BASE_ACCOUNT_CREATION_COST = fromMAS(0.001);
/* -------------------------------------------------------------------------- */
/*                                 GAS LIMIT                                 */
/* -------------------------------------------------------------------------- */
export const MAX_GAS_EXECUTE_SC = 3980167295n;
export const MAX_GAS_DEPLOYMENT = MAX_GAS_EXECUTE_SC;
export const MAX_GAS_CALL = 4294167295n;
export const MIN_GAS_CALL = 2100000n;
/* -------------------------------------------------------------------------- */
/*                                   NETWORK                                  */
/* -------------------------------------------------------------------------- */
export const MAINNET = 'MainNet';
export const BUILDNET = 'BuildNet';
export const SECURENET = 'SecureNet';
export const LABNET = 'LabNet';
export const SANDBOX = 'Sandbox';
export const MAINNET_CHAIN_ID = 77658377n;
export const BUILDNET_CHAIN_ID = 77658366n;
export const SECURENET_CHAIN_ID = 77658383n;
export const LABNET_CHAIN_ID = 77658376n;
export const SANDBOX_CHAIN_ID = 77n;
// Adjusted: Use the values of the constants as keys
export const CHAIN_ID_TO_NETWORK_NAME = {
    [MAINNET_CHAIN_ID.toString()]: MAINNET,
    [BUILDNET_CHAIN_ID.toString()]: BUILDNET,
    [SECURENET_CHAIN_ID.toString()]: SECURENET,
    [LABNET_CHAIN_ID.toString()]: LABNET,
    [SANDBOX_CHAIN_ID.toString()]: SANDBOX,
}; // type is inferred as the specific, unchangeable structure
// NETWORK_NAME_TO_CHAIN_ID mapping remains the same
export const CHAIN_ID = {
    [MAINNET]: MAINNET_CHAIN_ID,
    [BUILDNET]: BUILDNET_CHAIN_ID,
    [SECURENET]: SECURENET_CHAIN_ID,
    [LABNET]: LABNET_CHAIN_ID,
    [SANDBOX]: SANDBOX_CHAIN_ID,
}; // type is inferred as the specific, unchangeable structure
export var DefaultProviderUrls;
(function (DefaultProviderUrls) {
    DefaultProviderUrls["MAINNET"] = "https://mainnet.massa.net/api/v2";
    DefaultProviderUrls["TESTNET"] = "https://test.massa.net/api/v2";
    DefaultProviderUrls["BUILDNET"] = "https://buildnet.massa.net/api/v2";
    DefaultProviderUrls["LABNET"] = "https://labnet.massa.net/api/v2";
    DefaultProviderUrls["LOCALNET"] = "http://127.0.0.1";
})(DefaultProviderUrls || (DefaultProviderUrls = {}));
export const CHAIN_ID_RPC_URL_MAP = {
    [MAINNET_CHAIN_ID.toString()]: DefaultProviderUrls.MAINNET,
    [BUILDNET_CHAIN_ID.toString()]: DefaultProviderUrls.BUILDNET,
    [SANDBOX_CHAIN_ID.toString()]: DefaultProviderUrls.LOCALNET,
}; // type is inferred as the specific, unchangeable structure
export const CHAIN_ID_DNS_ADDRESS_MAP = {
    [MAINNET_CHAIN_ID.toString()]: 'AS1q5hUfxLXNXLKsYQVXZLK7MPUZcWaNZZsK7e9QzqhGdAgLpUGT',
    [BUILDNET_CHAIN_ID.toString()]: 'AS12qKAVjU1nr66JSkQ6N4Lqu4iwuVc6rAbRTrxFoynPrPdP1sj3G',
}; // type is inferred as the specific, unchangeable structure
